### PR TITLE
The Basics to Making your Web Application Work Offline

### LINKED ISSUE #4437: The Basics to Making your Web Application Work Offline
#### Proposed title of article
The Basics to Making your Web Application Work Offline

#### Proposed article introduction
The number of web applications has increased tremendously in the recent past. These applications require internet connectivity, which partly explains why there is a growing demand for internet services to gain access to them. 

However, there are instances when users experience internet connection losses, which disrupts the use of online applications. Some of the causes of internet connection losses include poor weather, power outages and poor internet coverage. Such instances have increased the need for applications that can work in offline mode. 

This article takes you through the basics to making web applications work offline. It will also go through the advantages of having an offline mode and the tools for storing data offline. 

#### Key takeaways 
1. Offline mode explained 
2. Benefits of having an application that works offline
3. Deciding offline features for your application
4. Choosing offline mode technologies
5. Offline data storage (tools for mobile and web)
6. Data sync for the applications
7. Conclusion

#### Article quality
This article provides handy tips that can help developers improve the user experience of their applications. It is unique because it presents information that is not readily available in other online resources. 

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by artmuth**
This fixes #4437

### CODE REVIEW

File: content/articles/the-basics-to-making-your-web-application-work-offline/index.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+The number of web applications has increased tremendously in the recent past. These applications require internet connectivity, which partly explains why there is a growing demand for internet services to gain access to them. 
+
+However, there are instances when users experience internet connection losses, which disrupts the use of online applications. Some of the causes of internet connection losses include poor weather, power outages, and poor internet coverage. Such instances have increased the need for applications that can work in offline mode.
```
Comment (mbaabuones) [inline content/articles/the-basics-to-making-your-web-application-work-offline/index.md:3]:
```suggestion
However, there are instances when users experience lost internet connection, which disrupts the use of online applications. Some of the causes of internet connection losses include poor weather, power outages, and poor internet coverage. Such instances have increased the need for applications that can work in offline mode.

File: content/articles/the-basics-to-making-your-web-application-work-offline/index.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  
+### Implementing offline mode
+The implementation of offline mode in web applications depends on the following operations:
+- Offline data storage
+- Data synchronization
+  
+#### Offline data storage
+When building a web application that supports offline mode, data is stored in the browser. The following are the main data storage tools employed:
+- **IndexedDB:** This is also termed as Indexed Database API. It is used for making data available to the client-side of the application in offline mode. Your application should not allow HTTPS calls to avoid failure in offline mode. 
+- **Service workers:** These tools work as proxies for the client-side of applications. When clients make network requests, ServiceWorkers intercept them and send responses to the clients.  Offline mode does not depend on networks. In this mode, downloaded data is stored in an offline cache. Service workers fetch cached responses from the offline cache and send them to the clients.
```
Comment (mbaabuones) [inline content/articles/the-basics-to-making-your-web-application-work-offline/index.md:44]: use 'service workers consistently in your article

File: content/articles/the-basics-to-making-your-web-application-work-offline/index.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  
+The following diagram shows the lifecycle of the service workers. 
+
+![Service Workers](/engineering-education/the-basics-to-making-your-web-application-work-offline/service-workers.png)
+
+[Image Source](https://yalantis.com/uploads/ckeditor/pictures/4093/service-workers-for-a-web-app.png)
+
+#### Data sync for the applications
+When the user is in offline mode due to a lost internet connection, he can still perform some actions. The IndexedDB stores these user actions. The corresponding data is synchronized with the server when the network connection resumes so that any pending jobs can be processed. It is important to set the triggers of data synchronization. The developers should set a reasonable sync frequency that optimizes battery usage.
```
Comment (mbaabuones) [inline content/articles/the-basics-to-making-your-web-application-work-offline/index.md:53]: put IndexedDB in bold

**Review by mbaabuones — CHANGES_REQUESTED** [review]
@artmuth Great article! Your article is well structured and easy to understand. Please work on the minor suggestions highlighted in the article and ping me when you are done.

**Comment by WanjaMIKE** [comment]
Please correct the plagiarism in highlighted areas. Thanks

[New Scan 12_43 PM Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7512805/New.Scan.12_43.PM.Copyleaks.Report.pdf)

**Comment by artmuth** [comment]
> Please correct the plagiarism in highlighted areas. Thanks
> 
> [New Scan 12_43 PM Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7512805/New.Scan.12_43.PM.Copyleaks.Report.pdf)

@WanjaMIKE Done.

**Comment by artmuth** [comment]
@mbaabuones I have finished revising the article.

**Comment by hectorkambow** [comment]
Jumping on this one

**Comment by hectorkambow** [comment]
👍 overall nice work - very detailed and straight forward tutorial 
 
Please review and suggestions/changes made by your peer reviewer and/or myself during a final review.

**Pull request merged**